src/jvm/main/org/jetbrains/kotlinx/lincheck/util/UnsafeHolder.kt

```diff
@@ -79,4 +79,81 @@ internal fun getArrayElementOffsetViaUnsafe(arr: Any, index: Int): Long {
+    return when (field.type) {
+        Boolean::class.javaPrimitiveType    -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putBoolean(obj, field, value as Boolean) }
+        Byte::class.javaPrimitiveType       -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putByte(obj, field, value as Byte) }
+        Char::class.javaPrimitiveType       -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putChar(obj, field, value as Char) }
+        Short::class.javaPrimitiveType      -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putShort(obj, field, value as Short) }
+        Int::class.javaPrimitiveType        -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putInt(obj, field, value as Int) }
+        Long::class.javaPrimitiveType       -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putLong(obj, field, value as Long) }
+        Double::class.javaPrimitiveType     -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putDouble(obj, field, value as Double) }
+        Float::class.javaPrimitiveType      -> writeFieldViaUnsafe(obj, field, value) { obj, field, value -> putFloat(obj, field, value as Float) }
+        else                                -> error("No more types expected")
+    }
+}
+
+internal fun writeArrayElementViaUnsafe(arr: Any, index: Int, value: Any?): Any? {
+    val offset = getArrayElementOffsetViaUnsafe(arr, index)
+    val componentType = arr::class.java.componentType
+
+    if (!componentType.isPrimitive) {
+        return UnsafeHolder.UNSAFE.putObject(arr, offset, value)
+    }
+
+    return when (componentType) {
+        Boolean::class.javaPrimitiveType    -> UnsafeHolder.UNSAFE.putBoolean(arr, offset, value as Boolean)
+        Byte::class.javaPrimitiveType       -> UnsafeHolder.UNSAFE.putByte(arr, offset, value as Byte)
+        Char::class.javaPrimitiveType       -> UnsafeHolder.UNSAFE.putChar(arr, offset, value as Char)
+        Short::class.javaPrimitiveType      -> UnsafeHolder.UNSAFE.putShort(arr, offset, value as Short)
+        Int::class.javaPrimitiveType        -> UnsafeHolder.UNSAFE.putInt(arr, offset, value as Int)
+        Long::class.javaPrimitiveType       -> UnsafeHolder.UNSAFE.putLong(arr, offset, value as Long)
+        Double::class.javaPrimitiveType     -> UnsafeHolder.UNSAFE.putDouble(arr, offset, value as Double)
+        Float::class.javaPrimitiveType      -> UnsafeHolder.UNSAFE.putFloat(arr, offset, value as Float)
+        else                                -> error("No more primitive types expected")
+    }
+}
+
+@Suppress("DEPRECATION")
+internal fun getFieldOffsetViaUnsafe(field: Field): Long {
+    return if (Modifier.isStatic(field.modifiers)) {
+        UnsafeHolder.UNSAFE.staticFieldOffset(field)
+    }
+    else {
+        UnsafeHolder.UNSAFE.objectFieldOffset(field)
+    }
+}
+
+@Suppress("DEPRECATION")
+internal fun findFieldNameByOffset(targetType: Class<*>, offset: Long): String? {
```

Comment:
Shouldn't be `findFieldNameByOffsetViaUnsafe` ?

Reply:
done